### PR TITLE
Add BadPacketsV

### PR DESCRIPTION
**Pull request by ManInMyVan**
Adds a check to validate `PLAYER_BLOCK_PLACEMENT` with `OTHER` face (used as `USE_ITEM` for pre-1.9)
| tested on        | Paper 1.8.8 | Paper 1.12.2 | Paper 1.16.5 |
|----------------|-------------|--------------|--------------|
| Forge 1.7.10   | ✔️              |                     |                     |
| Forge 1.12.2   | ✔️              | ✔️               |                     |
| Forge 1.8.9     | ✔️              | ✔️               |                     |
| Vanilla 1.16.5  | ✔️              |                     | ✔️               |
| Vanilla 1.17.1  | ✔️              |                     |                     |
| Fabric 1.20.4   | ✔️              |                     |                     |

### CODE REVIEW

**Comment by MachineBreaker** [comment]
Reuse the position methods result to make it easier to maintain

**Comment by ManInMyVan** [comment]
> Reuse the position methods result

like this?
```diff
-                if (packet.getBlockPosition().getX() != -1
-                        || packet.getBlockPosition().getY() != expectedY
-                        || packet.getBlockPosition().getZ() != -1
-                        || packet.getCursorPosition().getX() != 0
-                        || packet.getCursorPosition().getY() != 0
-                        || packet.getCursorPosition().getZ() != 0
+                if (!packet.getBlockPosition().equals(new Vector3i(-1, expectedY, -1))
+                        || !packet.getCursorPosition().equals(Vector3f.zero())

**Comment by MachineBreaker** [comment]
Whatever returns getBlockPosition() -> Object blockPosition = getBlockPosition() (change Object by the return type of getBlockPosition)

**Comment by ManInMyVan** [comment]
> Whatever returns getBlockPosition() -> Object blockPosition = getBlockPosition() (change Object by the return type of getBlockPosition)

So this?
```java
final Vector3i pos = packet.getBlockPosition();
final Vector3f cur = packet.getCursorPosition();
if (pos.getX() != -1
         || pos.getY() != expectedY
         || pos.getZ() != -1 
         || cur.getX() != 0 
         || cur.getY() != 0 
         || cur.getZ() != 0 
         || !packet.getItemStack().isPresent()
) flagAndAlert();
```

**Comment by MachineBreaker** [comment]
> > Whatever returns getBlockPosition() -> Object blockPosition = getBlockPosition() (change Object by the return type of getBlockPosition)
> 
> So this?
> 
> ```java
> final Vector3i pos = packet.getBlockPosition();
> final Vector3f cur = packet.getCursorPosition();
> if (pos.getX() != -1
>          || pos.getY() != expectedY
>          || pos.getZ() != -1 
>          || cur.getX() != 0 
>          || cur.getY() != 0 
>          || cur.getZ() != 0 
>          || !packet.getItemStack().isPresent()
> ) flagAndAlert();
> ```

Ye

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                // except y gets wrapped?
+                final int expectedY = player.getClientVersion().isNewerThanOrEquals(ClientVersion.V_1_8) ? 4095 : 255;
+
+                if (packet.getBlockPosition().getX() != -1
+                        || packet.getBlockPosition().getY() != expectedY
+                        || packet.getBlockPosition().getZ() != -1
+                        || packet.getCursorPosition().getX() != 0
+                        || packet.getCursorPosition().getY() != 0
+                        || packet.getCursorPosition().getZ() != 0
+                        || !packet.getItemStack().isPresent() // never sent when not holding anything
```
Comment (SamB440) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java]: Would it be good to check the player's inventory as well?

Comment (ManInMyVan) [reply src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java]: Just tested this, and `player.getInventory().getHeldItem()` had too much desync when spamming right click with blocks.

Comment (ManInMyVan) [reply src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java]: It should also probably be it's own check

**Comment by z7087** [comment]
False with forge 1.12.2 and clumsy 1000ms
I think it should exempt 1.9+ players (on itemstack check)

**Comment by ManInMyVan** [comment]
> False with forge 1.12.2 and clumsy 1000ms
> I think it should exempt 1.9+ players (on itemstack check)

is this reproducible on 1.20?

**Comment by z7087** [comment]
I think yes but I didn't test
[HandItemCache.java#L41](https://github.com/ViaVersion/ViaVersion/blob/fe0d30df782501c1bb81ad4c4be8c8194fe061ee/bukkit-legacy/src/main/java/com/viaversion/viaversion/bukkit/listeners/protocol1_9to1_8/HandItemCache.java#L41)
Via set itemstack from bukkit in the player_block_placement packet(with 1.9+ client 1.8- server)

**Comment by ManInMyVan** [comment]
> I think yes but I didn't test

~~Can you test it?~~

> [HandItemCache.java#L41](https://github.com/ViaVersion/ViaVersion/blob/fe0d30df782501c1bb81ad4c4be8c8194fe061ee/bukkit-legacy/src/main/java/com/viaversion/viaversion/bukkit/listeners/protocol1_9to1_8/HandItemCache.java#L41)
> Via set itemstack from bukkit in the player_block_placement packet(with 1.9+ client 1.8- server)

I'll exempt 1.9+ for itemstack

~~Actually I think we can exempt 1.9+ completely. (https://github.com/ViaVersion/ViaVersion/blob/master/common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/WorldPackets.java#L281-L334)~~ Can't because #1375 is all versions

**Comment by z7087** [comment]
图片
reproducible on spigot 1.8.8 with fabric 1.20.2

**Comment by AoElite** [comment]
@ManInMyVan Can you make it experimental

**Pull request merged**